libbpf-rs/src/map.rs

```diff
@@ -241,12 +238,180 @@ impl Map {
+                fd: MapFd::Borrowed(fd),
+                name,
+                ty,
+                key_size,
+                value_size,
+            },
+            ptr,
+        })
+    }
+
+    /// Returns whether map is pinned or not flag
+    pub fn is_pinned(&self) -> bool {
+        unsafe { libbpf_sys::bpf_map__is_pinned(self.ptr.as_ptr()) }
+    }
+
+    /// Returns the pin_path if the map is pinned, otherwise, None is returned
+    pub fn get_pin_path(&self) -> Option<&OsStr> {
+        let path_ptr = unsafe { libbpf_sys::bpf_map__pin_path(self.ptr.as_ptr()) };
+        if path_ptr.is_null() {
+            // means map is not pinned
+            return None;
+        }
+        let path_c_str = unsafe { CStr::from_ptr(path_ptr) };
+        Some(OsStr::from_bytes(path_c_str.to_bytes()))
+    }
+
+    /// [Pin](https://facebookmicrosites.github.io/bpf/blog/2018/08/31/object-lifetime.html#bpffs)
+    /// this map to bpffs.
+    pub fn pin<P: AsRef<Path>>(&mut self, path: P) -> Result<()> {
+        let path_c = util::path_to_cstring(path)?;
+        let path_ptr = path_c.as_ptr();
+
+        let ret = unsafe { libbpf_sys::bpf_map__pin(self.ptr.as_ptr(), path_ptr) };
+        util::parse_ret(ret)
+    }
+
+    /// [Unpin](https://facebookmicrosites.github.io/bpf/blog/2018/08/31/object-lifetime.html#bpffs)
+    /// this map from bpffs.
+    pub fn unpin<P: AsRef<Path>>(&mut self, path: P) -> Result<()> {
+        let path_c = util::path_to_cstring(path)?;
+        let path_ptr = path_c.as_ptr();
+        let ret = unsafe { libbpf_sys::bpf_map__unpin(self.ptr.as_ptr(), path_ptr) };
+        util::parse_ret(ret)
+    }
+
+    /// Returns an iterator over keys in this map
+    ///
+    /// Note that if the map is not stable (stable meaning no updates or deletes) during iteration,
+    /// iteration can skip keys, restart from the beginning, or duplicate keys. In other words,
+    /// iteration becomes unpredictable.
+    pub fn keys(&self) -> MapKeyIter {
+        MapKeyIter::new(self, self.key_size())
+    }
+
+    /// Attach a struct ops map
+    pub fn attach_struct_ops(&self) -> Result<Link> {
+        if self.map_type() != MapType::StructOps {
+            return Err(Error::InvalidInput(format!(
+                "Invalid map type ({}) for attach_struct_ops()",
+                self.map_type(),
+            )));
+        }
+
+        util::create_bpf_entity_checked(|| unsafe {
+            libbpf_sys::bpf_map__attach_struct_ops(self.ptr.as_ptr())
+        })
+        .map(|ptr| unsafe {
+            // SAFETY: the pointer came from libbpf and has been checked for errors
+            Link::new(ptr)
+        })
+    }
+
+    /// Retrieve the underlying [`libbpf_sys::bpf_map`].
+    pub fn as_libbpf_bpf_map_ptr(&self) -> NonNull<libbpf_sys::bpf_map> {
+        self.ptr
+    }
+
+    /// Create a new [`MapHandle`] from this map.
+    pub fn to_handle(&self) -> Result<MapHandle> {
+        self.handle.try_clone()
+    }
```

Comment:
Let's remove this method? Now that we deref into a `MapHandle` this is seems like unnecessary duplication. It's arguably a bit confusing that `<Map>.try_clone()` produces a `MapHandle`, at least from a naming perspective, but logically it does make sense.

Reply:
Oh yeah right, now you can call `try_clone` directly. How about renaming `try_clone` to `new_handle`/`clone_handle`/`to_new_handle`/`to_handle` or something like that? Then it would make sense in both call sites.

Reply:
I actually kind of like `try_clone`. `to_*` methods probably should not be fallible. I suppose `try_clone_handle` would be okay -- the way I see it, we either end up with duplicate information when called on a `MapHandle` or insufficient context when called on ` Map`.

The one thing we could do is change the `MapHandle::try_clone` signature to:
```rust
pub fn try_clone(this: &MapHandle) -> Result<Self>
```

That's similar to what various pointer-style objects in the standard library do (e.g., https://doc.rust-lang.org/std/rc/struct.Rc.html#method.try_unwrap). It would mean that callers have to use `MapHandle::try_clone(&<map-handle>)`. Perhaps that would be the least confusing? I don't have a strong preference either way.

Reply:
I guess this is one of the reasons why the docs advise to be careful with `Deref`-implementations: They can be confusing.

I personally like Rust's `&self` syntax more, but I see why the stdlib does it like this. It's definitely the most explicit way, so probably the least confusing.